File: src/federal_register.py
from requests import get

class FederalRegister:
	def __init__(self) -> None:
		self.api = "https://www.federalregister.gov/api/v1"
		self.headers = {
			"user-agent": "Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/103.0.0.0 Safari/537.36"
		}

	def get_document(
			self,
			document_number: str) -> dict:
		return get(
			f"{self.api}/documents/{document_number}.json",
			headers=self.headers).json()

	def get_documents(
			self,
			document_numbers: str) -> dict:
		return get(
			f"{self.api}/documents/{document_numbers}.json",
			headers=self.headers).json()

	def search_all_documents(
			self,
			fields: str = None,
			per_page: int = 20,
			page: int = 1,
			order: str = None,
			conditions_term: str = None) -> dict:
		url = f"{self.api}/documents.json?per_page={per_page}&page={page}"
		if fields:
			url += f"&fields[]={fields}"
		if order:
			url += f"&order={order}"
		if conditions_term:
			url += f"&conditions[term]={conditions_term}"
		return get(
			url, headers=self.headers).json()

	def get_documents_grouped_by_facet(
			self,
			facet: str = "daily",
			conditions_term: str = None) -> dict:
		url = f"{self.api}/documents/facets/{facet}"
		if conditions_term:
			url += f"?conditions[term]={conditions_term}"
		return get(
			url, headers=self.headers).json()


	def get_public_inspection_document(
			self,
			document_number: str) -> dict:
		return get(
			f"{self.api}/public-inspection-documents/{document_number}.json",
			headers=self.headers).json()


	def get_public_inspection_documents(
			self,
			document_numbers: str) -> dict:
		return get(
			f"{self.api}/public-inspection-documents/{document_numbers}.json",
			headers=self.headers).json()

	def get_current_public_inspection_documents(self) -> dict:
		return get(
			f"{self.api}/public-inspection-documents/current.json",
			headers=self.headers).json()

	def search_all_public_inspection_documents(
			self,
			conditions_available_on: str,
			fields: str = None,
			per_page: int = 20,
			page: int = 1,
			order: str = None,
			conditions_term: str = None) -> dict:
		url = f"{self.api}/public-inspection-documents.json?per_page={per_page}&page={page}&conditions[available_on]={conditions_available_on}"
		if fields:
			url += f"&fields[]={fields}"
		if order:
			url += f"&order={order}"
		if conditions_term:
			url += f"&conditions[term]={conditions_term}"
		return get(
			url, headers=self.headers).json()

	def get_all_agency_details(self) -> dict:
		return get(
			f"{self.api}/agencies",
			headers=self.headers).json()

	def get_particular_agency_details(
			self,
			slug: str,
			federal_register_id: int) -> dict:
		return get(
			f"{self.api}/agencies/{slug}?id={federal_register_id}",
			headers=self.headers).json()

File: src/federal_register.js
class FederalRegister {
	constructor() {
		this.api = "https://www.federalregister.gov/api/v1"
		this.headers = {
			"User-Agent": "Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/103.0.0.0 Safari/537.36"
		}
	}

	async getDocument(documentNumber) {
		const response = await fetch(
			`${this.api}/documents/${documentNumber}.json`, {
				method: "GET",
				headers: this.headers
			})
		return response.json()
	}

	async searchAllDocuments(perPage = 20, page = 1, fields = null, order = null, conditionsTerm = null) {
		let url = `${this.api}/documents.json?per_page=${perPage}&page=${page}`
		if (fields) {
			url += `&fields=${fields}`
		}
		if (order) {
			url += `&order=${order}`
		}
		if (conditionsTerm) {
			url += `&conditions[term]=${conditionsTerm}`
		}
		const response = await fetch(
			url, {
				method: "GET",
				headers: this.headers
			})
		return response.json()
	}

	async getDocumentsGroupedByFaceT(faceT = "daily", conditionsTerm = null) {
		let url = `${this.api}/documents/facets/${faceT}`
		if (conditionsTerm) {
			url += `?conditions[term]=${conditionsTerm}`
		}
		const response = await fetch(
			url, {
				method: "GET",
				headers: this.headers
			})
		return response.json()
	}

	async getPublicInspectionDocument(documentNumber) {
		const response = await fetch(
			`${this.api}/public-inspection-documents/${documentNumber}.json`, {
				method: "GET",
				headers: this.headers
			})
		return response.json()
	}

	async getCurrentPublicInspectionDocument() {
		const response = await fetch(
			`${this.api}/public-inspection-documents/current.json`, {
				method: "GET",
				headers: this.headers
			})
		return response.json()
	}

	async searchAllPublicInspectionDocuments(conditionsAvailableOn, perPage = 20, page = 1, fields = null, order = null, conditionsTerm = null) {
		let url = `${this.api}/public-inspection-documents.json?per_page==${perPage}&page=${page}&conditions[available_on]=${conditionsAvailableOn}`
		if (fields) {
			url += `&fields=${fields}`
		}
		if (order) {
			url += `&order=${order}`
		}
		if (conditionsTerm) {
			url += `&conditions[term]=${conditionsTerm}`
		}
		const response = await fetch(
			url, {
				method: "GET",
				headers: this.headers
			})
		return response.json()
	}

	async getAgencies() {
		const response = await fetch(
			`${this.api}/agencies`, {
				method: "GET",
				headers: this.headers
			})
		return response.json()
	}

	async getParticularAgencies(slug, federalRegisterId) {
		const response = await fetch(
			`${this.api}/agencies/${slug}?id=${federalRegisterId}`, {
				method: "GET",
				headers: this.headers
			})
		return response.json()
	}
}

module.exports = {FederalRegister}
